server/src/rooms/lobby/tournament/index.ts

```diff
@@ -27,19 +33,56 @@ export class TournamentLobbyRoom extends LobbyRoom<TournamentLobbyRoomState> {
   async trySendInvitations() {
     // only try to send invitations if there are enough players and we are not in the middle
     // of trying to put groups into a game
-    if (this.processingClients) {
-      return;
+    const release = await this.mutex.acquire();
+    try {
+      this.formAndInviteGroups();
+    } finally {
+      release();
     }
-    this.processingClients = true;
-    const clients = this.state.getEligibleClients();
-    if (clients.length >= 5) {
-      const shuffledClients = _.shuffle(clients);
-      while (shuffledClients.length >= 5) {
-        const group = shuffledClients.splice(0, 5);
+  }
+
+  async formAndInviteGroups() {
+    while (this.state.queue.length >= this.groupSize) {
+      const groupId = this.formGroup();
+      const group = this.state.getPendingGroup(groupId);
+      if (group) {
         await this.sendGroupInvitations(group);
+        this.setGroupAcceptanceTimeout(groupId);
+      }
+    }
+  }
+
+  setGroupAcceptanceTimeout(groupId: string) {
+    // set a timeout for 10 seconds to accept the invitation, otherwise assume something
+    // bad happened and clear the group
+    const timeout = setTimeout(() => {
+      if (!this.state.allGroupClientsAccepted(groupId)) {
+        this.state.resetPendingGroup(groupId);
       }
+      this.groupTimeouts.delete(groupId);
+    }, 1000 * 10);
+    this.groupTimeouts.set(groupId, timeout);
+  }
+
+  clearGroupAcceptanceTimeout(groupId: string) {
+    const timeout = this.groupTimeouts.get(groupId);
+    if (timeout) {
+      clearTimeout(timeout);
+      this.groupTimeouts.delete(groupId);
+    }
+  }
+
+  formGroup(): string {
+    const group = [];
+    const groupId = uuidv4();
+    _.shuffle(this.state.queue);
```

Comment:
would it be more straightforward to shuffle once and drain all valid groups at once and then leave the unlucky leftovers in the queue?

Reply:
Yeah, the queue isn't actually managed correctly, I'll look at this when doing so